[Discover] Add field types in-product help

andreadelrio opened the pull request:
## Summary

This is a follow-up to the work started in ON week.

- Adds a help button next to the `Filter by type` dropdown. This button displays a popover with information about field types.

<img width="515" alt="image" src="https://user-images.githubusercontent.com/4016496/162301894-503b3d48-c113-4c9b-8691-fa9d618f82ec.png">

- Field types list taken from here: 
https://github.com/elastic/kibana/blob/2f468993d425ca89e0037920df8f8aaedf61f382/src/plugins/discover/public/application/main/components/sidebar/lib/get_field_type_name.ts 

### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
~- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials~
- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [ ] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
~- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)~
- [x] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [x] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)


### Risk Matrix

Delete this section if it is not applicable to this PR.

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| Multiple Spaces&mdash;unexpected behavior in non-default Kibana Space. | Low | High | Integration tests will verify that all features are still supported in non-default Kibana Space and when user switches between spaces. |
| Multiple nodes&mdash;Elasticsearch polling might have race conditions when multiple Kibana nodes are polling for the same tasks. | High | Low | Tasks are idempotent, so executing them multiple times will not result in logical error, but will degrade performance. To test for this case we add plenty of unit tests around this logic and document manual testing procedure. |
| Code should gracefully handle cases when feature X or plugin Y are disabled. | Medium | High | Unit tests will verify that any feature flag or plugin combination still results in our service operational. |
| [See more potential risk examples](https://github.com/elastic/kibana/blob/main/RISK_MATRIX.mdx) |


### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

kertal commented:
@elasticmachine merge upstream

cchaos commented:
I 😍 this extra in-page documentation. I wonder if we can keep it around instead of allowing it to be dismissed? For instance, what if the `Filter by type` input had an `append` that triggers this popover? That way, ES continues to add support for more field types, the user will always be able to find the full list. Maybe even add a popoer footer to a full docs page?
<img width="561" alt="Screen Shot 2022-03-02 at 10 09 44 AM" src="https://user-images.githubusercontent.com/549577/156389504-84beeb25-ae6e-4461-a7ab-163da310fd26.png">

andreadelrio commented:
> I 😍 this extra in-page documentation. I wonder if we can keep it around instead of allowing it to be dismissed? For instance, what if the `Filter by type` input had an `append` that triggers this popover? That way, as ES continues to add support for more field types, the user will always be able to find the full list. Maybe even add a popoer footer to a full docs page? <img alt="Screen Shot 2022-03-02 at 10 09 44 AM" width="561" src="https://user-images.githubusercontent.com/549577/156389504-84beeb25-ae6e-4461-a7ab-163da310fd26.png">

I think this is a good idea! I was a bit worried about users dismissing the callout and not being able to access this info afterward. The filter dropdown currently uses **EuiFilterButton** which doesn't support an append. I've used **EuiButtonIcon** instead which I think works well too.

<img width="714" alt="image" src="https://user-images.githubusercontent.com/4016496/156453518-59ba0d58-9747-4bc3-9ea9-f0e1fd65e1a2.png">

@gchaps what should we include in the footer?

cchaos commented:
☹️ These ragged lines are breaking my designer heart.
<img width="714" alt="156453518-59ba0d58-9747-4bc3-9ea9-f0e1fd65e1a2" src="https://user-images.githubusercontent.com/549577/156454199-209f9096-caea-4874-85b8-f93f74b3e6e8.png">

Instead of an `append` you could just turn it into a filter group and add another EuiFilterButton like so: https://codesandbox.io/s/youthful-bohr-uq8tnh?file=/demo.js
<img width="324" alt="Screen Shot 2022-03-02 at 16 50 15 PM" src="https://user-images.githubusercontent.com/549577/156455474-2ef34e9c-9030-4d1e-b2e4-c9d48c3dd907.png">

andreadelrio commented:
@cchaos Thanks for the feedback. This is what I came up with. I had to resort to an EUI mixin (`euiFocusBackground`) to add a fill to the help button. Let me know if you know of a better way to achieve that.

<img width="727" alt="image" src="https://user-images.githubusercontent.com/4016496/156471125-0efd8883-1ad4-4652-9b88-24519bbb3a2e.png">

miukimiu commented:
@andreadelrio @cchaos I just want to share some guidelines that we've been working on with @gchaps: [Figma project](https://www.figma.com/file/tEKAoonjs9hN2pgZJkoVYG/In-product-help?node-id=0%3A1).

It is still a work in progress but the idea is to have some guidelines to ensure consistency in the way we show product help across our products.

- In our work in progress, the icon to this use case (documentation that opens in a popover) would be the `documentation` icon. [Figma frame](https://www.figma.com/file/tEKAoonjs9hN2pgZJkoVYG/In-product-help?node-id=78%3A84148).
- Because it's clickable it would show in a primary color.

Of course, these are only ideas we have... Still WIP. So feel free to ignore them and we can revisit them later on. 

It would also be great to also have your thoughts on the guidelines. So feel free to add comments on the [Figma project](https://www.figma.com/file/tEKAoonjs9hN2pgZJkoVYG/In-product-help?node-id=0%3A1).

andreadelrio commented:
> @andreadelrio @cchaos I just want to share some guidelines that we've been working on with @gchaps: [Figma project](https://www.figma.com/file/tEKAoonjs9hN2pgZJkoVYG/In-product-help?node-id=0%3A1).
> 
> It is still a work in progress but the idea is to have some guidelines to ensure consistency in the way we show product help across our products.
> 
> * In our work in progress, the icon to this use case (documentation that opens in a popover) would be the `documentation` icon. [Figma frame](https://www.figma.com/file/tEKAoonjs9hN2pgZJkoVYG/In-product-help?node-id=78%3A84148).
> * Because it's clickable it would show in a primary color.
> 
> Of course, these are only ideas we have... Still WIP. So feel free to ignore them and we can revisit them later on.
> 
> It would also be great to also have your thoughts on the guidelines. So feel free to add comments on the [Figma project](https://www.figma.com/file/tEKAoonjs9hN2pgZJkoVYG/In-product-help?node-id=0%3A1).

This is great @miukimiu ! Thanks for bringing it up. I actually think this could also fall under the `Field description - long text` category as we're not explaining how a feature works but providing info on the `field types`. This would be a new case to consider adding to our guidelines under that category: using an append.

On a separate note and following up on the background color which I added to try to mimic the behavior of **EuiFormLayout**'s append. I wanted to point out that because of the way this is currently built when the EuiFilterButton on the left receives focus it gets a gray background that matches the color of the "append". I don't love how that looks. @cchaos Do you think this makes it worth considering adding proper prepend/append support for **EuiFilterGroup**? Especially considering we're developing guidelines for this type of in-product help.

![Screen Recording 2022-03-03 at 12 08 33 PM](https://user-images.githubusercontent.com/4016496/156645057-53df0f85-e286-4f76-8576-45398b1a2249.gif)

cchaos commented on src/plugins/discover/public/application/main/components/sidebar/discover_field_search.scss:
```diff
@@ -2,6 +2,10 @@
 }
+
+.dscFieldTypesHelp__button {
+  @include euiFocusBackground($euiColorDarkShade);
```
Can you further explain what you mean by "I don't love how that looks" when using the gray?

If you're trying to match the append background color, it's not using the focus color but:
```suggestion
  background-color: $euiFormInputGroupLabelBackground;
  min-width: 0 !important; // Reduce width of icon-only button
```

I'd also recommend removing the min-width on the button so that it's only as wide as it needs to be (hence that second line ^^)

gchaps commented on src/plugins/discover/public/application/main/components/sidebar/discover_field_search.tsx:
```diff
@@ -80,13 +83,71 @@ export function DiscoverFieldSearch({ onChange, value, types }: Props) {
+      id: 0,
+      dataType: 'text',
+      type: 'tokenString',
+      description: 'Full text such as the body of an email or a product description.',
```
```suggestion
      description: 'Full text, such as the body of an email or a product description.',
```

gchaps commented on src/plugins/discover/public/application/main/components/sidebar/discover_field_search.tsx:
```diff
@@ -80,13 +83,71 @@ export function DiscoverFieldSearch({ onChange, value, types }: Props) {
+      dataType: 'keyword',
+      type: 'tokenKeyword',
+      description:
+        'Structured content such as an ID, email address, hostname, status code, or tag.',
```
```suggestion
        'Structured content, such as an ID, email address, hostname, status code, or tag.',
```

gchaps commented on src/plugins/discover/public/application/main/components/sidebar/discover_field_search.tsx:
```diff
@@ -80,13 +83,71 @@ export function DiscoverFieldSearch({ onChange, value, types }: Props) {
+      id: 3,
+      dataType: 'date',
+      type: 'tokenDate',
+      description: 'A date string or the number of seconds or milliseconds since 1/1/1970.',
```
```suggestion
      description: 'A date string, or the number of seconds or milliseconds since 1/1/1970.',
```

cchaos commented on src/plugins/discover/public/application/main/components/sidebar/discover_field_search.tsx:
```diff
@@ -272,30 +344,51 @@ export function DiscoverFieldSearch({ onChange, value, types }: Props) {
       <EuiSpacer size="xs" />
-      <EuiOutsideClickDetector onOutsideClick={() => {}} isDisabled={!isPopoverOpen}>
-        <EuiFilterGroup className="dscFieldSearch__filterWrapper">
+      <EuiFlexItem>
+        <EuiFilterGroup fullWidth>
+          <EuiOutsideClickDetector onOutsideClick={() => {}} isDisabled={!isPopoverOpen}>
+            <EuiPopover
```
I think there's some extraneous, very old leftover wrappers here that aren't necessary anymore
```suggestion
        <EuiFilterGroup fullWidth>
          <EuiPopover
```

cchaos commented on src/plugins/discover/public/application/main/components/sidebar/discover_field_search.tsx:
```diff
@@ -257,6 +318,17 @@ export function DiscoverFieldSearch({ onChange, value, types }: Props) {
+      className="dscFieldTypesHelp__button"
+      aria-label="Filter type help"
+    >
+      <EuiIcon type="questionInCircle" title="Filter type help" />
```
```suggestion
      <EuiIcon type="iInCircle" color="primary" title="Filter type help" />
```

gchaps reviewed (commented):
Is this the complete list of field types?  I thought there were more.

As for the footer, I agree that we should include a link to the ES docs.

cchaos reviewed (commented):
Great reference @miukimiu , thank you! We really should get those finalized.

I agree with @andreadelrio here though, that this isn't a "How this feature works" but more like a cheatsheet. I see it falling more under the category of `Field description: Long content with formatting` so probably just need to change this to an `iInCircled` icon.

andreadelrio commented:
@kertal I pushed some changes to hopefully clean up the code for the table and added a separate file to keep the description texts (now i18n ready). Do you think you can help me add pagination to it? 6 rows per page would be ideal and we don't need the "Rows per page" button.

kertal commented:
> @kertal I pushed some changes to hopefully clean up the code for the table and added a separate file to keep the description texts (now i18n ready). Do you think you can help me add pagination to it? 6 rows per page would be ideal and we don't need the "Rows per page" button.

Thx for this great work! For sure I can help, but more likely next week. There are also some "unknown" fields like "date_range" we should explain (after researching what they're doing)

andreadelrio commented:
> > @kertal I pushed some changes to hopefully clean up the code for the table and added a separate file to keep the description texts (now i18n ready). Do you think you can help me add pagination to it? 6 rows per page would be ideal and we don't need the "Rows per page" button.
> 
> Thx for this great work! For sure I can help, but more likely next week. There are also some "unknown" fields like "date_range" we should explain (after researching what they're doing)

@kertal awesome thanks! @gchaps is working on the remaining descriptions

kertal commented:
@elasticmachine merge upstream

kertal commented:
dear @Dmitriynj could you help here, adding pagination, I fear I won't be able this week

gchaps reviewed (commented):
Here are the remaining field types.  I couldn't find a description of `conflict`, `string`, and `unknown`. @kertal can you help?

Also, how should we order these:  alphabetically or by most popular?  I think by most popular.


<meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-be6abcf6-7fff-86cf-b8ad-43ca4759c9c1"><div dir="ltr" style="margin-left:0pt;" align="left">

Type | Description
-- | --
boolean | True and false values.
conflict |  Values in the field that are of different types.  Resolve in Management > Data Views.
date | Date or the number of seconds or milliseconds since 1/1/1970.
geo_point | Latitude and longitude points.
geo_shape | Complex shapes, such as polygons.
ip | IPv4 and IPv6 addresses.
murmur3 | Field that computes and stores hashes of values.
number | Long, integer, short, byte, double, and float values.
source | Original JSON body provided during indexing.
~~string~~ |  
text | Full text, such as the body of an email or a product description
keyword | Structured content, such as an ID, email address, hostname, status code, or tag.
nested | JSON object that preserves the relationship between its subfields.
version | Software versions. Supports Semantic Versioning precedence rules.
unknown |  No information exists on the field mapping. The field might not be not indexed.

</div></b>

kertal commented:
> Here are the remaining field types. I couldn't find a description of `conflict`, `string`, and `unknown`. @kertal can you help?

**string** seams to be legacy:

https://www.elastic.co/de/blog/strings-are-dead-long-live-strings

**conflict**

you have an index pattern index*

field A in indexA is keyword
field A in indexB is text 

this causes a conflict in the mapping

**unknown**
We simply have no information about the mapping of a field, it could be unmapped


> Also, how should we order these: alphabetically or by most popular? I think by most popular.

popular is a bit legacy here in Discover, alphabetically is old school and rock solid, tested for a few thousand years 😄

ryankeairns commented:
Try it out on the test deployment, looking great!

andreadelrio commented:
@gchaps seems like we're missing the definitions for `ip_range` and `date_range`.

andreadelrio commented:
@kertal is `_source` the same as `source`? 
I'm seeing this comment for `source` in a related file:
https://github.com/elastic/kibana/blob/655d692945cf5bae5aa951d06fe20b5829db2c39/src/plugins/discover/public/application/main/components/sidebar/lib/get_field_type_name.ts#L46

@gchaps provided the following for `source`: 

source | Original JSON body provided during indexing.
-- | --

Just want to double-check if we're ok to use the definition Gail provided for `source` for `_source`

gchaps commented:
I added the descriptions for `ip_range` and `date_range` to the table above.

@kertal, can you comment on `source`?

For the link, how about 

> Learn more about [field types](https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-types.html).

kertal commented:
Sorry for the late reply, `_source` should be the same like `source`. Both should not be offered in the sidebar, so I think we could remove it. BTW when testing it, I noticed a problem in the responsive view
<img width="528" alt="Bildschirmfoto 2022-03-28 um 13 33 20" src="https://user-images.githubusercontent.com/463851/160390164-2e892c1b-90a3-4c98-ac88-dec446af5023.png">
Last question, it seems the area to trigger the help takes a bit much horizontal screen space, would be nice to align width to the data  picker ... even knowing that the picker soon will be gone
<img width="343" alt="Discover_-_Elastic" src="https://user-images.githubusercontent.com/463851/160390430-d1267041-fdbf-4c5c-b02c-3d7326e4dd8e.png">

andreadelrio commented:
@kertal I removed _source from the popover and I also fixed the width of the button and the responsive view. I would need some help with making the link to `Learn more` work. I tried to follow the pattern we follow for the Document Explorer callout but it seems like I'm generating the link wrong.

There are also some failing tests I'll likely need help with. I'll reach out to @Dmitriynj

andreadelrio commented:
@elasticmachine merge upstream

cchaos commented on src/plugins/discover/public/application/main/components/sidebar/discover_field_search.tsx:
```diff
@@ -257,6 +316,18 @@ export function DiscoverFieldSearch({ onChange, value, types }: Props) {
+      onClick={onHelpClick}
+      data-test-subj="fieldTypesHelpButton"
+      className="dscFieldTypesHelp__button"
+      aria-label="Filter type help"
+    >
+      <EuiIcon type="iInCircle" color="primary" title="Filter type help" />
```
Both `aria-label` and `title` need to be i18n-ized

cchaos commented on src/plugins/discover/public/application/main/components/sidebar/discover_field_search.tsx:
```diff
@@ -294,8 +365,58 @@ export function DiscoverFieldSearch({ onChange, value, types }: Props) {
+            {presentFieldTypes.length > FIELD_TYPES_PER_PAGE && (
+              <EuiFlexGroup justifyContent="flexEnd" gutterSize="none" responsive={false}>
+                <EuiFlexItem grow={false}>
+                  <EuiPagination
```
I **really** don't think pagination is necessary here, there's only like 10 fields that have descriptions right now and can often leave stragglers (1 row) on a single page. My suggestion is just to wrap the table in an overflowing container once it reaches a certain height. I'll setup a PR against this to showcase what I mean but essentially it would behave like this: 

https://user-images.githubusercontent.com/549577/161828833-3397c362-87f1-4202-8a54-5a7bb6bf172c.mp4

cchaos replied:
PR4U: https://github.com/andreadelrio/kibana/pull/7

cchaos commented on src/plugins/discover/public/application/main/components/sidebar/discover_field_search.tsx:
```diff
@@ -294,8 +365,58 @@ export function DiscoverFieldSearch({ onChange, value, types }: Props) {
+              </EuiFlexGroup>
+            )}
+            <EuiPopoverFooter paddingSize="s">
+              Learn more about&nbsp;
```
i18n

cchaos reviewed (commented):
💟 I love the permanent placement of this and how it's only showing field types that exist in the selected Data View. I have some nit-picky design suggestions that I'll pass to you in a PR.

andreadelrio commented on src/plugins/discover/public/application/main/components/sidebar/lib/get_field_type_description.ts:
```diff
@@ -0,0 +1,72 @@
+        defaultMessage: 'Software versions. Supports Semantic Versioning precedence rules.',
+      });
+    default:
+      return i18n.translate('discover.fieldNameDescription.unknownField', {
```
@gchaps No, this is a fallback in case there's a field type with a missing description.

andreadelrio commented:
@kertal please let me know if it's possible to have links inside `i18n.translate` like @gchaps is asking.
@cchaos thanks for your comments and PR! It looks a lot better. I felt something off with the footer and your solution is perfect. I merged your PR and addressed the rest of your feedback.

Dmitriynj commented:
> @kertal please let me know if it's possible to have links inside `i18n.translate` like @gchaps is asking. 

Yes, there is an example [here](https://github.com/elastic/kibana/blob/60e06c39970f4ce518d6afb167878c9b6ea9407c/src/plugins/discover/server/ui_settings.ts#L174).

```js
i18n.translate('discover.advancedSettings.disableDocumentExplorerDescription', {
  defaultMessage:
    'To use the new {documentExplorerDocs} instead of the classic view, turn off this option. ' +
    'The Document Explorer offers better data sorting, resizable columns, and a full screen view.',
  values: {
    documentExplorerDocs:
      `<a href=${docLinks.links.discover.documentExplorer}
        target="_blank" rel="noopener">` +
      i18n.translate('discover.advancedSettings.documentExplorerLinkText', {
        defaultMessage: 'Document Explorer',
      }) +
      '</a>',
  },
});
```

kertal commented on src/plugins/discover/public/application/main/components/sidebar/lib/get_field_type_description.ts:
```diff
@@ -0,0 +1,72 @@
+      return i18n.translate('discover.fieldNameDescription.numberField', {
+        defaultMessage: 'Long, integer, short, byte, double, and float values.',
+      });
+    case 'string':
```
no, we can remove `string`  👍 

https://www.elastic.co/de/blog/strings-are-dead-long-live-strings

kertal commented on src/plugins/discover/public/application/main/components/sidebar/discover_sidebar.tsx:
```diff
@@ -196,16 +196,20 @@ export function DiscoverSidebarComponent({
-        if (result.indexOf(field.type) === -1) {
-          result.push(field.type);
+        if (field.type !== '_source') {
+          dataViewFieldTypes.add(field.type);
```
This is complicated and we should simplify later on. 

So Es typed fields `text` and `keyword` are mapped to the Kibana field type string. So we need to split up this type. I've added code here so it should work, but feel free to extract the extraction of the actual used type to a separate function, so it's better readable code. Apart from that it's now really nice and functional 👍 
```suggestion
          // If it's a string type, we want to distinguish between keyword and text
          // For this purpose we need the ES type
          const type =
            field.type === 'string' &&
            field.esTypes &&
            ['keyword', 'text'].includes(field.esTypes[0])
              ? field.esTypes?.[0]
              : field.type;
          // _id and _index would map to string, that's why we don't add the string type here
          if (type && type !== 'string') {
            dataViewFieldTypes.add(type);
          }
```

kertal commented:
> > @kertal please let me know if it's possible to have links inside `i18n.translate` like @gchaps is asking.
Just wanted to add to @Dmitriynj suggestion, there's another way to do it, using react components in the values property:
https://github.com/elastic/kibana/blob/d0c06b0112c97eb83f67a9375a4d92b57ef03b25/src/plugins/discover/public/application/main/components/document_explorer_callout/document_explorer_update_callout.tsx#L70-L80

FYI @jughosta , sharing i18n knowledge that's not obvious

cchaos commented on src/plugins/discover/public/application/main/components/sidebar/discover_field_search.tsx:
```diff
@@ -80,13 +99,54 @@ export function DiscoverFieldSearch({ onChange, value, types }: Props) {
 
+  const { docLinks } = useDiscoverServices();
+
+  const { curPageIndex, pageSize, totalPages, startIndex, changePageIndex } = usePager({
+    initialPageSize: FIELD_TYPES_PER_PAGE,
+    totalItems: presentFieldTypes.length,
+  });
```
Since the pagination component is removed now, there's a lot of logic/variables you don't need either.

kertal commented:
@elasticmachine merge upstream

andreadelrio commented on src/plugins/discover/public/application/main/components/sidebar/lib/get_field_type_description.ts:
```diff
@@ -0,0 +1,72 @@
+      });
+    case 'version':
+      return i18n.translate('discover.fieldNameDescription.versionField', {
+        defaultMessage: 'Software versions. Supports Semantic Versioning precedence rules.',
```
Added it!

andreadelrio commented:
@elasticmachine merge upstream

gchaps reviewed (approved):
LGTM!!

cchaos reviewed (approved):
Looks great!
